lib/Dialect/AIE/Transforms/AIEGenerateColumnControlOverlay.cpp

```diff
@@ -0,0 +1,255 @@
+                                        sourceBundle, sourceChannel);
+    builder.create<AIE::PacketDestOp>(builder.getUnknownLoc(), dest, destBundle,
+                                      destChannel);
+    builder.create<AIE::EndOp>(builder.getUnknownLoc());
+    return pktFlow;
+  }
+
+  // Create packet flows per col which moves control packets to and from shim
+  // dma
+  void generatePacketFlowsForControl(OpBuilder builder, DeviceOp device,
+                                     const AIETargetModel &targetModel,
+                                     TileOp shimTile, WireBundle shimWireBundle,
+                                     SmallVector<int> availableShimChans,
+                                     SmallVector<AIE::TileOp> ctrlTiles,
+                                     WireBundle ctrlWireBundle,
+                                     int coreOrMemChanId,
+                                     int unusedPacketIdFrom, bool isShimMM2S) {
+    // Create packet flows
+    SmallVector<int> thresholdsToNextShimChannel;
+    for (int i = 1; i < (int)availableShimChans.size() + 1; i++)
+      thresholdsToNextShimChannel.push_back(ctrlTiles.size() /
+                                            availableShimChans.size() * i);
+    int ctrlPktFlowID = unusedPacketIdFrom;
+    int shimChanIdx = 0;
+    for (int i = 0; i < (int)ctrlTiles.size(); i++) {
+      if (ctrlTiles[i]->hasAttr("controller_id"))
+        ctrlPktFlowID =
+            (int)ctrlTiles[i]
+                ->getAttrOfType<AIE::PacketInfoAttr>("controller_id")
+                .getPktId();
+      builder.setInsertionPointToEnd(device.getBody());
+      if (isShimMM2S)
+        (void)createPacketFlowOp(builder, ctrlPktFlowID, shimTile,
+                                 shimWireBundle,
+                                 availableShimChans[shimChanIdx], ctrlTiles[i],
+                                 ctrlWireBundle, coreOrMemChanId);
+      else
+        (void)createPacketFlowOp(builder, ctrlPktFlowID, ctrlTiles[i],
+                                 ctrlWireBundle, coreOrMemChanId, shimTile,
+                                 shimWireBundle,
+                                 availableShimChans[shimChanIdx]);
+      if (i >= thresholdsToNextShimChannel[shimChanIdx]) {
+        shimChanIdx++;
+        ctrlPktFlowID = unusedPacketIdFrom;
+      }
+    }
+  }
+
+  // Get a vector of shim channels not reserved by any circuit-switched aie.flow
+  // op
+  SmallVector<int> getAvailableShimChans(DeviceOp device, int numShimChans,
+                                         TileOp shimTile,
+                                         WireBundle shimWireBundle,
+                                         bool isShimMM2S) {
+    SmallVector<int> availableShimChans;
+    for (int i = 0; i < numShimChans; i++) {
+      bool isShimChanReservedByAIEFlowOp = false;
+      device.walk([&](AIE::FlowOp fOp) {
+        if (isShimMM2S && fOp.getSource() == shimTile &&
```

Comment:
Could also iterate over the users of `shimTile`? To avoid walking over all ops in the device `numShimChans` times.

Reply:
Great idea. Code optimized as suggested.